### DISCUSSION TITLE
VTK Slice Disappear After Some User Interaction

### DISCUSSION THREAD

**Discussion by wyattnapier** [Kitware/trame#540] [Trame/VTK] [answered]
Hi! I am having a similar issue with my volumes disappearing as @maflancer in [discussion post #539](https://github.com/Kitware/trame/discussions/539). However, I have not yet deployed my application though, so that isn't part of my issue. When I am using files with fewer blocks I can interact with the slices for longer (changing slice index or plane), whereas files with more blocks only permit a few interactions. This may be because my approach to re-rendering the slice actors is inefficient, but the memory consumption of my application appears to be constant according to Windows Task Manager. This is what my output looks like when functioning properly:
![Screenshot 2024-06-13 150348](https://github.com/Kitware/trame/assets/117198964/ae139096-2049-412f-88dd-80ac682b1c4b)
As described in the previous post, I also don't get any errors in the console or the terminal, so I'm not sure what the cause of this issue is. Unfortunately, my code is a bit messier than the code in the previous post, but I'll respond quickly to any questions about it.

Here is the function that handles the change in slice plane which then triggers a few more functions to change the VOI:
```
@state.change("slice_plane")
def read_slice_plane_change(slice_plane, **kwargs):
    if slice_plane == "I":
        state.I_slice = 1
    elif slice_plane == "J":
        state.J_slice = 1
    elif slice_plane == "K":
        state.K_slice = 1
    ctrl.view_update()
@state.change("I_slice")
def update_I_slice_VOI(I_slice, **kwargs):
    update_slices("I", I_slice, state.slices_visibility)
    ctrl.view_update()
@state.change("J_slice")
def update_J_slice_VOI(J_slice, **kwargs):
    update_slices("J", J_slice, state.slices_visibility)
    ctrl.view_update()
@state.change("K_slice")
def update_K_slice_VOI(K_slice, **kwargs):
    update_slices("K", K_slice, state.slices_visibility)
    ctrl.view_update()
```

Then the two key functions for actually changing the VOI are:
```
### wrapper function to call build slices with updated VOI and the prior info for extractGrid
def update_slices(slice, sliceIndex, visBool):
    for loopIndex in range(len(combined_grid)):
        if grid_slice_actor[loopIndex]:
            renderer.RemoveActor(grid_slice_actor[loopIndex])

        ### vtkCellDataToPointData
        cellDataToPointData = vtkCellDataToPointData()
        cellDataToPointData.SetInputData(combined_grid[loopIndex])

        build_slice(cellDataToPointData, loopIndex, slice, sliceIndex, visBool)

### specific function to build the slices
def build_slice(cellDataToPointData, loopIndex, slice, sliceIndex, visBool):
    # adjust voi for slice
    if slice == "I":
        input_VOI = (sliceIndex-1, sliceIndex, 0, BLOCK_IJK[loopIndex][1], 0, BLOCK_IJK[loopIndex][2])
    elif slice == "J":
        input_VOI = (0, BLOCK_IJK[loopIndex][0], sliceIndex-1, sliceIndex, 0, BLOCK_IJK[loopIndex][2])
    elif slice == "K":
        input_VOI = (0, BLOCK_IJK[loopIndex][0], 0, BLOCK_IJK[loopIndex][1], sliceIndex-1, sliceIndex)
    else:
        assert False, "invalid slice"

    ### vtkExtractGrid to get the slices
    extractGrid = vtkExtractGrid()
    extractGrid.SetInputConnection(cellDataToPointData.GetOutputPort())
    extractGrid.SetVOI(input_VOI)
    print("new voi:", extractGrid.GetVOI())
    extractGrid.SetIncludeBoundary(True) 

    sliceDataSetSurfaceFilter = vtkDataSetSurfaceFilter()
    sliceDataSetSurfaceFilter.SetInputConnection(extractGrid.GetOutputPort())

    grid_slice_mapper = vtkPolyDataMapper()
    grid_slice_mapper.SetInputConnection(extractGrid.GetOutputPort())
    grid_slice_actor[loopIndex] = vtkActor()
    grid_slice_actor[loopIndex].SetMapper(grid_slice_mapper)
    renderer.AddActor(grid_slice_actor[loopIndex])
    # grid_slice_actor[loopIndex].SetVisibility(visBool)
```
Finally, the following is my GUI for the slices card:
```
def slices_card():
    with ui_card(title="Slices", ui_name="slices"):
        vuetify.VCheckbox(
            label="Show Filter",
            v_model=("slices_visibility", VISIBILITY_ARR[2]),
            on_icon="mdi-checkbox-marked",
            off_icon="mdi-checkbox-blank-outline",
            classes="mx-1",
            hide_details=True,
            dense=True,
        )
        vuetify.VSelect(
            v_model=("slice_plane", None),
            items=(
                "planes",
                [
                    {"text": "I", "value": "I"},
                    {"text": "J", "value": "J"},
                    {"text": "K", "value": "K"},
                ],
            ),
            label="Slice Plane",
            hide_details=True,
            dense=True,
            outlined=True,
            classes="pt-1",
        )
        with vuetify.VRow(v_if="slice_plane == 'I'", classes="pt-2", dense=True):
            vuetify.VSlider(
                # gridline wdith
                v_model=("I_slice", 0),
                min=1,
                max=MIN_VOI[0],
                step=1,
                label="I Slice Selector",
                classes="mt-1",
                hide_details=True,
                dense=True,
            )
            with vuetify.VBtn(icon=True, click=reset_I_slice):
                vuetify.VIcon("mdi-restore")
            vuetify.VTextField(
                # display I_slice index
                v_model=("I_slice", 0),
                label="I Slice",
                classes="mt-1",
                hide_details=True,
                dense=False,
                readonly=False,  # make the text field read-only
                suffix="/" + str(MIN_VOI[0])
            )
        with vuetify.VRow(v_else_if="slice_plane == 'J'", classes="pt-2", dense=True):
            vuetify.VSlider(
                v_model=("J_slice", 0),
                min=1,
                max=MIN_VOI[1],
                step=1,
                label="J Slice Selector",
                classes="mt-1",
                hide_details=True,
                dense=True,
            )
            with vuetify.VBtn(icon=True, click=reset_J_slice):
                vuetify.VIcon("mdi-restore")
            vuetify.VTextField(
                v_model=("J_slice", 1),
                label="J Slice",
                classes="mt-1",
                hide_details=True,
                dense=False,
                readonly=False,
                suffix="/" + str(MIN_VOI[1])
            )
        with vuetify.VRow(v_else_if="slice_plane == 'K'", classes="pt-2", dense=True):
            vuetify.VSlider(
                v_model=("K_slice", 1),
                min=1,
                max=MIN_VOI[2],
                step=1,
                label="K Slice Selector",
                classes="mt-1",
                hide_details=True,
                dense=True,
            )
            with vuetify.VBtn(icon=True, click=reset_K_slice):
                vuetify.VIcon("mdi-restore")
            vuetify.VTextField(
                v_model=("K_slice", 1),
                label="K Slice",
                classes="mt-1",
                hide_details=True,
                dense=False,
                readonly=False,
                suffix="/" + str(MIN_VOI[2])
            )
```
For reference, most of my code is based off of the [example application in the trame tutorial](url) except my data is a mutli-block dataset.

EDIT: I am using `view = vtk.VtkRemoteLocalView(renderWindow, namespace="view", mode="local", interactive_ratio=1)`, but when I tried to switch to solely vtkLocalView it functioned even more poorly.

**Comment by jourdain** [comment, marked as answer]
So for VtkLocalView and the Local part of the VtkRemoteLocalView might get confused with the Add/RemoveActor. Try to play with visibility of the actor instead. Otherwise, if you use the RemoteView it will just work.

**Reply by wyattnapier** [reply]
I tried changing to VtkRemoteView so that I could reduce the number of actors I would need and none of the slices appeared, although the other filters still worked, even when I hardcoded the visibility of the slice actors to be True.

If I were to just change visibility of the actors instead of adding and removing them then I would end up with an actor for a slice across every index on the I plane, as well as the J plane, and the K plane. That is just for one block and at full scale I could have hundreds of blocks. Would trame be able to support this volume or would you recommend a different approach for managing visibility?

**Reply by jourdain** [reply]
Why do you need 1 actor per slice? Can't you just update filter/mapper to adjust the plane? 

Also if something is not working with VtkRemoteView, that means your vtk code is wrong.

**Reply by wyattnapier** [reply]
It seems that by not adding any more actors and thus not having to remove them it did fix the issue when working with VtkLocalView. However, I still haven't been able to fix the issue for the VtkRemoteView.  Both my slices and the gridlines from my cube axes drop away as shown below. Do you have any ideas about what is wrong with my vtk code?
![image](https://github.com/Kitware/trame/assets/117198964/881f8a4d-5a5f-4c23-958f-f1b2c2e54bf9)

**Reply by jourdain** [reply]
Probably because of invalid mapper for the data type you are dealing with.

**Reply by wyattnapier** [reply]
You were right, thank you. It was very similar to a [prior discussion post](https://github.com/Kitware/trame/discussions/476#discussioncomment-8766082). I'm not sure if the background gridlines for the cube axes are able to show up in VtkRemoteView because in the [trame tutorial example](https://kitware.github.io/trame/guide/tutorial/application.html#vtk_pipeline-id) when you run solution.py the same issue occurs.

**Reply by jourdain** [reply]
It is just settings on the vtk object. The VtkRemoteView truly show what vtk is doing. The local one, you get some translation that can change what gets rendered.
